File: Silver III/BOJ_1654/BOJ_1654.cpp
#include <algorithm>
#include <iostream>
#include <vector>

#define ll long long

using namespace std;

int n;
ll k;
vector<int> info;

int BS(ll low, ll high)
{
    int ret = 0;
    ll mid;
    int size = info.size();

    // cout << "low, high : " << low << ", " << high << '\n';

    while (low <= high) {
        mid = (low + high) >> 1;
        // cout << "mid : " << mid << '\n';

        ll temp = 0;
        for (int i = 0; i < size; ++i) {
            temp += (ll)(info[i] / mid);
        }

        if (k <= temp) {      // 현재 개수가 원하는 개수를 뛰어 넘었다면
            if (ret < mid) {  // 현재 길이가 더 크다면 갱신
                ret = mid;
            }
            low = mid + 1;
        }
        else {
            high = mid - 1;
        }
    }

    return ret;
}

int main()
{
    cin >> n >> k;
    info.resize(n);
    for (int i = 0; i < n; ++i) {
        cin >> info[i];
    }

    cout << BS(1LL, *max_element(info.begin(), info.end())) << '\n';

    return 0;
}

File: Silver III/BOJ_1654/BOJ_1654.py
import sys


def binarySearch(info, k):
    answer = 0

    low = 1
    high = max(info)
    # print(low, high)

    while low <= high:
        mid = (low + high) >> 1
        # print(mid)

        total = 0
        for row in info:
            total += row // mid

        if total >= k:
            if answer < mid:
                answer = mid
            low = mid + 1
        else:
            high = mid - 1

    return answer


if __name__ == "__main__":
    n, k = map(int, sys.stdin.readline().split())
    info = [int(sys.stdin.readline()) for _ in range(n)]

    # print(n, k, info)
    print(binarySearch(info, k))
